src/Lecture2.hs

```diff
@@ -268,7 +362,19 @@ data EvalError
 It returns either a successful evaluation result or an error.
 -}
 eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
+eval _ (Lit x) = Right x
+eval [] (Var varName) = Left (VariableNotFound varName)
+eval vars (Var varName) = maybeToEither $ lookup varName vars
+  where
+    maybeToEither:: Maybe Int -> Either EvalError Int
+    maybeToEither Nothing = Left (VariableNotFound varName)
+    maybeToEither (Just y) = Right y
+eval vars (Add expr1 expr2) = sumEither (eval vars expr1) (eval vars expr2)
+  where
+    sumEither:: Either EvalError Int -> Either EvalError Int -> Either EvalError Int
+    sumEither (Left (VariableNotFound x)) _ = Left (VariableNotFound x)
+    sumEither _ (Left (VariableNotFound x)) = Left (VariableNotFound x)
+    sumEither (Right x) (Right y) = Right (x + y)
```

Comment:
That's a great helper function! 👏🏻 
Once you learn about Monads, you'll be able to reduce this function to a single line ✂️